attributes/zzz_system_tuning.rb

```diff
@@ -0,0 +1,167 @@
+    ports += [node['hadoop']['core_site']['fs.defaultFS'].split(':')[2].to_i]
+  else
+    # Default port for hdfs:// is 8020
+    ports += [8020]
+  end
+else
+  ports += [8020]
+end
+
+# These are all address:port pairs with no protocol prefix specified
+# eg. 0.0.0.0:50070
+addr_ports = {
+  'dfs.namenode.http-address' => 50_070,
+  'dfs.namenode.https-address' => 50_470,
+  'dfs.datanode.http.address' => 50_075,
+  'dfs.datanode.https.address' => 50_475,
+  'dfs.namenode.secondary.http-address' => 50_090,
+  'dfs.namenode.secondary.https-address' => 50_091,
+  'dfs.datanode.address' => 50_010,
+  'dfs.datanode.ipc.address' => 50_020,
+  'dfs.namenode.backup.address' => 50_100,
+  'dfs.namenode.backup.http-address' => 50_105,
+  'dfs.journalnode.rpc-address' => 8485,
+  'dfs.journalnode.http-address' => 8480,
+  'dfs.journalnode.https-address' => 8481
+}
+
+addr_ports.each do |k,v|
+  if node['hadoop']['hdfs_site'].key?(k)
+    ports += [node['hadoop']['hdfs_site'][k].split(':')[1].to_i]
+  else
+    ports += [v]
+  end
+end
+
+###
+# MapReduce
+###
+
+if node['hadoop'].key?('mapred_site') && node['hadoop']['mapred_site'].key?('mapreduce.jobhistory.webapp.address')
+  ports += [node['hadoop']['mapred_site']['mapreduce.jobhistory.webapp.address'].split(':')[1].to_i]
+else
+  ports += [19_888]
+end
+
+###
+# YARN
+###
+
+addr_ports = {
+  'yarn.resourcemanager.scheduler.address' => 8030,
+  'yarn.resourcemanager.resource-tracker.address' => 8031,
+  'yarn.resourcemanager.address' => 8032,
+  'yarn.resourcemanager.admin.address' => 8033,
+  'yarn.resourcemanager.webapp.address' => 8088,
+  'yarn.resourcemanager.webapp.https.address' => 8090,
+  'yarn.nodemanager.localizer.address' => 8040,
+  'yarn.nodemanager.webapp.address' => 8042,
+  'yarn.timeline-service.address' => 10_200,
+  'yarn.timeline-service.webapp.address' => 8188,
+  'yarn.timeline-service.webapp.https.address' => 8190
+}
+
+addr_ports.each do |k,v|
+  if node['hadoop']['yarn_site'].key?('k')
+    ports += [node['hadoop']['yarn_site'][k].split(':')[1].to_i]
+  else
+    ports += [v]
+  end
+end
+
+###
+# HBase
+###
+
+addr_ports = {
+  'hbase.master.port' => 60_000,
+  'hbase.master.info.port' => 60_010,
+  'hbase.regionserver.port' => 60_020,
+  'hbase.regionserver.info.port' => 60_030,
+  'hbase.rest.port' => 8080,
+  'hbase.rest.info.port' => 8085,
+  'hbase.thrift.port' => 9090,
+  'hbase.thrift.info.port' => 9095
+}
+
+addr_ports.each do |k,v|
+  if node['hbase']['hbase_site'].key?(k)
+    ports += [node['hbase']['hbase_site'][k].to_i]
+  else
+    ports += [v]
+  end
+end
+
+###
+# Hive
+###
+
+if node['hive']['hive_site'].key?('hive.metastore.uris')
+  ms_arr = node['hive']['hive_site']['hive.metastore.uris'].split(',')
+  ms_arr.each do |ms|
+    # thrift://hostname:port
+    ports += [ms.split(':')[2].to_i]
+  end
+end
+
+###
+# ZooKeeper
+###
+
+if node['zookeeper'].key?('zoocfg') && node['zookeeper']['zoocfg'].key?('clientPort')
```

Comment:
we check if ``node['zookeeper'].key?('zoocfg')`` here, but not below on line 156. We should be consistent.  I think since there is a ``node['zookeeper']['zoocfg']`` default attribute, we can remove the check here.